### ISSUE TITLE
Feature request: Support KSP

### ISSUE THREAD

**Issue by Maksim-Novikov** [mapstruct/mapstruct#2522]
For Kotlin projects, Mapstruct currently makes use of KAPT. This is generally quite slow since it requires generating Java stubs before the annotation processing step.

KSP offers an alternative to this by making it easier for annotation processors like Mapstruct to more directly use Kotlin's compiler plugin infrastructure. Based on initial benchmarks, using KSP can speed up build times by as much as 2x.

https://github.com/google/ksp

**Comment by Wobbley** [comment]
Hey.

I was introduced to Mapstruct by a colleague last year, and it is pretty handy for converting between classes. Are there any plans on support KSP in the near future? We are looking at moving over to Java 17 as it is the new LTS, but Kotlin kapt does not work under Java 17 without adding a bunch of flags. So currently considering our options.

**Comment by kevinbos-routescanner** [comment]
Is there any progress/update on this? Would be a nice to have support 👍

**Comment by jakoss** [comment]
I think it's quite a work to support both kapt and ksp, but I thing there are libraries that allow to abstract most part of this. I think it's called X-Processing and it's used by room. Source here https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:room/room-compiler-processing/

**Comment by filiphr** [comment]
There hasn't been any progress on this currently and I am not sure when the MapStruct team is going to be able to look into this.

We do have some kind of a meta model in which we store the structure of the code we want to generate. However, we do have a lot of small nuances that are directly linked to our Freemarker templates (what we use under the hood to generate the java code). 

Perhaps it would be possible to perform more abstractions on our side to be able to extract things from KSP, but as @jakoss mentioned this would be a lot of work for us. 

For the time being we would like to focus on adding more features for the JavaAP and through that KAPT. The good thing with Kotlin is that it has a good compatibility with Java. If there

**Comment by StefanLobbenmeier** [comment]
@filiphr 

> If there

I think your comment was cut off here

**Comment by filiphr** [comment]
Oops you are right @StefanLobbenmeier.

I think what I wanted to say is that if there is someone from the community that would be interested into looking into this and  try to provide a good integration, we would be happy to help out.

**Comment by tokazio** [comment]
With the release of kotlin2.0 with it's New K2 compiler, kapt will be no more supported then mapstruct could not be used anymore.
Sad to see that the use of mapstruct avoir the use of the 2 times faster ksp and 2x times faster compiler...

**Comment by jakoss** [comment]
@tokazio https://github.com/s0nicyouth/kmapper

**Comment by tokazio** [comment]
In fact K2 is not supporting ksp at this time

**Comment by jakoss** [comment]
> In fact K2 is not supporting ksp at this time

It's the ksp that is not supporting k2 yet, but they will get there, no worries

**Comment by tokazio** [comment]
Yes no doubt on that but just a little rushed 😜
That gives Time to add ksp support to mapstruct

**Comment by jakoss** [comment]
I doubt that will happen at this point. There is little to no interest in the community as far i can tell. It seems like people like just writing Kotlin and mappers stays in the Java realm for some reason

**Comment by mcarleio** [comment]
I really liked using MapStruct with Java. But with Kotlin it was missing support for all the Kotlin sugar. I am developing (and just released) a KSP library which already provides the most core functionality. Give it a try: https://github.com/mcarleio/konvert 

Hint: It is of course not as powerful as MapStruct...

**Comment by filiphr** [comment]
I just wanted to let everyone know that KSP is on our radar and we are going to try and find the time to see how we can integrate MapStruct to work with it

**Comment by StefanLobbenmeierObjego** [comment]
https://github.com/androidx/androidx/tree/androidx-main/room/room-compiler-processing

is this an option to support ksp more easily?

**Comment by runningjustin** [comment]
Is there an update on KSP support? It's been about 11 months since the comment from @filiphr that it's on the roadmap.

**Comment by StefanLobbenmeierObjego** [comment]
To get a feel for how much effort this would be and maybe get the ball rolling I just spent some time looking into room compiler processing. From what I could see so far there is a lot of equivalents where migrating the code is just adding an X in front of it. 

On the other hand, there is one bigger difference in the ElementUtils and TypeUtils. Those are only available during javac annotation processing it seems, and there is an abstraction over it in XProcessingEnv, but its difficult to see how much mapstruct will need those / if all of those can be covered.

**Comment by StefanLobbenmeierObjego** [comment]
Found two more blockers:
- there is java 9 module support in mapstruct, but from what I could see not in room, so keeping this will be awkward. I think it could be possible to support modules when using javac and ignoring modules when using ksp, but I am not familiar with modules so this is hard to judge for me. Rough design for this would be to replace the checks for java 9 with (java9 && processingEnv instanceOf JavacProcessingEnv)

- there is another module https://github.com/mapstruct/tools-gem that would also need to changed to use the room types. It generates a lot of classes that use the java annotation classes and we would need to change this to the room classes

So I think this is now too big to be done by one individual 😄 gonna stop here for now, at least now I have a better idea of the difficulties that prevented this from being done earlier

**Comment by filiphr** [comment]
Thanks a lot for your efforts @StefanLobbenmeier. We are interacting a lot with the `javax.lang.model` classes. The difference in the API and the lack of familiarity with Kotlin is one of the biggest reasons why we haven't done this yet.

Whenever I find some more time I will try to play around with the PR you created @StefanLobbenmeier. One big question here is also about the generated code. Should we generate Java or Kotlin code? I would assume that people would expect to see Kotlin code instead of Java code for this.

Most likely we would first need to do certain refactoring in MapStruct that would allow us to more easily do KSP and maybe generate Kotlin code.

**Comment by StefanLobbenmeier** [comment]
> Whenever I find some more time I will try to play around with the PR you created

just keep in mind it does not compile yet, I was mostly naively going class to class seeing if the room replacements can work but there is still more to migrate.

> One big question here is also about the generated code. Should we generate Java or Kotlin code?

I assumed that the generated code can still be Java, but you are raising a good point. I think there is some expectation to be able to generate pure Kotlin, so it can be used for Kotlin multiplatform as well. 

 I use mapstruct in spring projects, so generating Java code is fine for me, it might just mean that the Java gradle plugin will be needed. So my expectation was that it generates a Java class that extends a Kotlin interface.

There is actually also a way to generate code with room, but I have not checked yet if it will always generate Kotlin with KSP and generate Java with Javac, but if so this would be a nice way to do this. Would create extra effort to replace Freeform with this so maybe this can be a follow up change?

**Comment by filiphr** [comment]
Good points @StefanLobbenmeier. As an initial step we could first support this by generating Java first through a kotlin interface. Once we have that iron out we can look into generating Kotlin as wel.

>Would create extra effort to replace freemarker with this so maybe this can be a follow up change?

I've had on my own list to replace Freemarker with JavaPoet. I've also seen that there is KotlinPoet that could be used to generate Kotlin. It would ba a longer road to do this, but as you said if we do it in steps it might be OK. We don't have to everything at once.

**Comment by tokazio** [comment]
I think MapStruct should have its own class model that can be a facade with a java implementation and a kotlin implementation

I've done that on some custom tools that i can implement with jandex or javac or ksp detection. Code generation is then separated (i used kotlinPoet and it was easy to use, i remember that it's the same API than javaPoet no ?)

I've looked at the current code and such an abstraction seems to need a lot of work
But this will separate the core MapStruct engine, the annotations processing and the code generation.

A first KSP support generating java code will be great enought 

It's a good news to targetting ksp

**Comment by meierw** [comment]
Would like this feature show up as well.

Currently Kapt is using functionality that will stop working in Gradle 9 -> https://youtrack.jetbrains.com/issue/KT-67838/Gradle-8.8-compatibility-Configuration.fileCollectionSpec-method-has-been-deprecated

It is unknown if JetBrains will or will not fix this, so it's unknown if `kapt` will continue to work in Gradle 9.  
I suspect they will fix it, cause popular libraries like Mapstruct still use `kapt`, but who knows. They haven't even prioritized the issue for a month :shrug:

**Comment by artemy-osipov** [comment]
JetBrains has fixed it in 2.0.20
https://youtrack.jetbrains.com/issue/KT-67888/Remove-usages-of-deprecated-Configuration.fileCollection-method

**Comment by meierw** [comment]
> JetBrains has fixed it in 2.0.20 https://youtrack.jetbrains.com/issue/KT-67888/Remove-usages-of-deprecated-Configuration.fileCollection-method

Ah. Indeed. Hadn't jumped up to v2 on my end yet.

**Comment by 0-don** [comment]
looking forward for this

**Comment by lingting** [comment]
If the workload is large, maybe you should list the plans and design specifications and put them in a new branch. This way, people who need them can better participate in the progress?

**Comment by hakan0xFF** [comment]
Are there any updates on this topic?

**Comment by honza-zidek** [comment]
@filiphr Can you implement this feature earlier than in version 1.7.0 ("No due date")? Something like 1.6.1 or so? :)

KAPT is **abandoned** by JetBrains (https://kotlinlang.org/docs/kapt.html):

> kapt is in maintenance mode. We are keeping it up-to-date with recent Kotlin and Java releases but have no plans to implement new features. Please use the [Kotlin Symbol Processing API (KSP)](https://kotlinlang.org/docs/ksp-overview.html) for annotation processing. [See the list of libraries supported by KSP](https://kotlinlang.org/docs/ksp-overview.html#supported-libraries).

The compiler warning is quite annoying
> [WARNING] Kapt currently doesn't support language version 2.0+. Falling back to 1.9.

and the impossibility to use advanced Kotlin features even more.

**We started considering using another framework for object mapping.**

**Comment by eriknyk** [comment]
4 years and we canot implement this feature, It would be healthier if they told us that this library is no longer maintained.?

**Comment by jakoss** [comment]
> 4 years and we canot implement this feature, It would be healthier if they told us that this library is no longer maintained.?

So is it "we" or "they"? 

This feature is clearly not a priority on a Java-first library and it's only fair. And it absolutely does not mean it's not maintained anymore.

Besides, there is a viable Kotlin counterpart mentioned in this thread. 

What is the point of your comment here?

**Comment by mcarleio** [comment]
> 4 years and we canot implement this feature, It would be healthier if they told us that this library is no longer maintained.?

As @jakoss suggested: have a look at https://github.com/mcarleio/konvert 

It does not have all features of MapStruct, but it certainly is very useful. Try it out for yourself and see if it suffices your use cases.

Hint: I am the author of Konvert.

**Comment by eriknyk** [comment]
> > 4 years and we canot implement this feature, It would be healthier if they told us that this library is no longer maintained.?
> 
> So is it "we" or "they"?
> 
> This feature is clearly not a priority on a Java-first library and it's only fair. And it absolutely does not mean it's not maintained anymore.
> 
> Besides, there is a viable Kotlin counterpart mentioned in this thread.
> 
> What is the point of your comment here?

I’m going to accept your invitation to leave this library.

**Comment by jakoss** [comment]
> > > 4 years and we canot implement this feature, It would be healthier if they told us that this library is no longer maintained.?
> > 
> > So is it "we" or "they"?
> > 
> > This feature is clearly not a priority on a Java-first library and it's only fair. And it absolutely does not mean it's not maintained anymore.
> > 
> > Besides, there is a viable Kotlin counterpart mentioned in this thread.
> > 
> > What is the point of your comment here?
> 
> I’m going to accept your invitation to leave this library.

It's not my library. And nobody forces you to use it 😂

**Comment by honza-zidek** [comment]
In a corporation with heavyweight bureaucratic processes and conservative coding rulebook, it's not always easy to change a library such "standard" as MapStruct. Hence would why many of us prefer if MapStruct accepted the challenge, accepted the fact that Kotlin is here and become more Kotlin compatible.

There are enough other problems with MapStruct in Kotlin, such as impossibility to use interface inheritance (and having to use abstract classes instead). But this is something we can live with.

**Comment by jakoss** [comment]
> Hence would why many of us prefer if MapStruct accepted the challenge

The best way to do that is to sponsor the library in a significant way. But shaming maintainers and calling the library unmaintained just because the feature you are after is not implemented - that is just not fair.

Look, i don't have a problem that you want a feature, i'd love it too. But this is still open source library. I'm not entitled to anything here. I can either wait or contribute. Complaining in open issues does nothing positive, ever.

**Comment by honza-zidek** [comment]
@jakoss 
1. It's clear and indisputable that authors of Open Source software have no moral duty to implement anything.
2. On other side, by "going public" you kind of become responsible for your creation. Especially if so many people depend on it.
3. I can understand someone's conclusion that if a library hasn't addressed an important issue for 4 years, that the library might easily look like unmaintained.

TL;DR: The open source author can implement or not implement whatever you want, the open source users can bitch about whatever they want :) By publishing your work you open it to judgment and criticism.

**Comment by filiphr** [comment]
Thanks for the support @jakoss.

@honza-zidek regarding your comments. 

> Hence would why many of us prefer if MapStruct accepted the challenge, accepted the fact that Kotlin is here and become more Kotlin compatible.

It is clear to us that Kotlin is here and that we need to make MapStruct more compatible with it. However, you have to understand that MapStruct is developed in our free time, there is no company behind it and we all do this in our free time.

**Comment by lostiniceland** [comment]
We happily use MapStruct in a Kotlin project with a little workaround, by using Java sourcefiles for Mapstruct :-)
You just have to place them in the same folder (like src/main/kotlin) and adjust the Java compiler.

There are some minor issues with JPA classes when (using the kotlin-jpa plugin) but this has nothing to do with Mapstruct.
```
@Embeddable
data class SomeClass @Default private constructor(val xxx: String.....)
```

The problem is that the jpa-plugin generates an empty constructor in order to work with JPA, and this is the one MapStruct sees by default.
You just need to create this `@Default` annotation yourself and Mapstruct will use the correct ctor.

Mapstruct is my most liked library because it takes so much pain away while still giving you fast feedback during compilation. Sure, I would like to use Mapstruct with Kotlin directly but this works great for us.

**Comment by AlexTrotsenko** [comment]
For someone who already now is looking for another alternative  - https://mappie.tech/getting-started/introduction/

I personally like simple dsl of this lib

**Comment by mcarleio** [comment]
> For someone who already now is looking for another alternative - https://mappie.tech/getting-started/introduction/
> 
> I personally like simple dsl of this lib

Oh, cool, did not know about that lib! After having a look I have to say I like the idiomatic Kotlin way 🔥

Nevertheless, I will continue developing Konvert, but I see a lot of potential in Mappie and am looking forward to see how both will evolve!

**Comment by realdadfish** [comment]
For mapstruct the issue of missing KSP support gets a little more problematic, as the new Android Gradle Plugin 9+ no longer supports `kapt` with its built-in Kotlin support. The workaround is to use the `com.android.legacy-kapt` plugin instead of `kotlin-kapt` then, but I'd say the days are counted.

**Comment by lostiniceland** [comment]
This basically means a re-implementation in KSP2. I've created a small AP and added Kotlin-KSP2 support by literally having two separate implementations. You probably can extract common functionality but it will be difficult to abstract over the differences in the type-system (Optional vs ?-Type, inline value types, primitives etc).

Next you need to setup a different testing approach. Java AP runs as part of the compiler, whereas KSP2 runs separately (before). Furthermore, currently KSP2 only supports Gradle officially (the Mill build tools has KSP2 support but not Maven).

At least in the latest release of KSP2 (2.3.x) they got rid of the dependency on the Kotlin compiler version. Before 2.3, the KSP version was tied to Kotlin version like `kolinversion_kspversion` (2.2.0_2.x.y).

So in general, not an easy feat.

**Comment by mcarleio** [comment]
For maven there are two inofficial KSP maven plugins: 
- https://github.com/mcarleio/ksp-maven-plugin
- https://github.com/kpavlov/ksp-maven-plugin

**Comment by filiphr** [comment]
I've had some time to play to think about this issue and how we can support this from MapStruct without a full rewrite. I've come up with the idea to generate a java mapper from a kotlin mapper.

e.g.

```kotlin
@Mapper
interface SimpleMapper {
    fun map(source: Source): Target
}
```

the KSP would generate soemthing like:

```java
@Mapper(implementationName = "SimpleMapperImpl")
interface SimpleMapperDriver extends SimpleMapper {
    Target map(Source source);
}
```

With this `Mapper.getMapper` would keep working as is and all the other functionality would just keep working.

It would be configured somehow like:

```gradle
dependencies {
    implementation("org.mapstruct:mapstruct")

    // KSP-side: translates Kotlin @Mapper interfaces to Java driver stubs.
    ksp("org.mapstruct:mapstruct-ksp")

    // javac-side: MapStruct's own annotation processor, which generates the impl of the driver.
    annotationProcessor("org.mapstruct:mapstruct-processor")
}
```

Would like to hear the opinion from the community about this approach?

**Comment by StefanLobbenmeier** [comment]
@filiphr interesting idea, but it’s pretty much exactly what kapt does. It also creates a Java class and then runs the Java annotation processors on it

I am not sure about the future of kapt, it was supposed to be deprecated but now the deprecation note disappeared

**Comment by XhstormR** [comment]
> I am not sure about the future of kapt, it was supposed to be deprecated but now the deprecation note disappeared

I think its because kSP cannot directly replace kapt. Kapt is Kotlin's compatibility solution with Java APT; KSP is Kotlin's native annotation/symbol handling solution for kotlin lang.

**Comment by e5l** [comment]
Hey! We're working on the kapt performance optimisations on the Kotlin side and it looks like we're able to achieve good numbers. 

We also going to have the benchmark suite to make sure kapt is fast and has no overhead, so looking for the samples and user projects. If anyone would like to participate or contribute their performance problems (including mapstruct), please let us know 🙂
